src/UserActivity.js

```diff
@@ -23,105 +23,160 @@ import Timer from './utils/Timer';
 // such as READ_MARKER_INVIEW_THRESHOLD_MS,
 // READ_MARKER_OUTOFVIEW_THRESHOLD_MS,
 // READ_RECEIPT_INTERVAL_MS in TimelinePanel
-const CURRENTLY_ACTIVE_THRESHOLD_MS = 2 * 60 * 1000;
+const CURRENTLY_ACTIVE_THRESHOLD_MS = 700;
+const CURRENTLY_PASSIVE_THRESHOLD_MS = 2 * 60 * 1000;
 
 /**
  * This class watches for user activity (moving the mouse or pressing a key)
  * and starts/stops attached timers while the user is active.
+ *
+ * There are two classes of 'active' a user can be: 'active' and 'passive':
+ * see doc on the userCurrently* functions for what these mean.
  */
-class UserActivity {
-    constructor() {
-        this._attachedTimers = [];
-        this._activityTimeout = new Timer(CURRENTLY_ACTIVE_THRESHOLD_MS);
+export default class UserActivity {
+    constructor(windowObj, documentObj) {
+        this._window = windowObj;
+        this._document = documentObj;
+
+        this._attachedTimersActive = [];
+        this._attachedTimersPassive = [];
+        this._activeTimeout = new Timer(CURRENTLY_ACTIVE_THRESHOLD_MS);
+        this._passiveTimeout = new Timer(CURRENTLY_PASSIVE_THRESHOLD_MS);
         this._onUserActivity = this._onUserActivity.bind(this);
         this._onWindowBlurred = this._onWindowBlurred.bind(this);
         this._onPageVisibilityChanged = this._onPageVisibilityChanged.bind(this);
         this.lastScreenX = 0;
         this.lastScreenY = 0;
     }
 
+    static sharedInstance() {
+        if (global.mxUserActivity === undefined) {
+            global.mxUserActivity = new UserActivity(window, document);
```

Comment:
So the reason we generally use global rather than module level variables / statics is that in node (and by extension, webpack etc) you get a different copy of the module for each module you import that module *from*: that is, we would define a module in react-sdk, then get a different copy of it when importing it from react-sdk to importing it from riot-web. This may not be so relevant now that almost all the code is in react-sdk, but as far as I know if you want one instance of something for the whole *app* rather than the calling module, this is the only way.

Reply:
Hmm, I really think that sounds like a bug somewhere in the chain, as the module system is not _supposed_ to work that way... 😖 For example, with Node:

```
> require('console').bob = true
true
> require('console').bob
true
```

It sounds like when you say "each module you import that module from", it's not about individual JS module boundaries, but instead something goes wrong between higher-level package boundaries (Riot vs. React SDK rather than `./foo.js` vs. `./bar.js`). I suspect this is connected to how we currently don't allow Webpack see into the SDKs at build time, and so it might be fixed once we do something like the purposed changes that have Webpack build Riot and the SDKs together in one step. (Let's do that soon!)

Anyway, I guess it's okay for now, since that's the world we have today. 😭

Reply:
I was just trying to make a test case to demonstrate but https://medium.com/@lazlojuly/are-node-js-modules-singletons-764ae97519af explains it well. Arguably if we end up in one of the situations that triggers this then we have bigger problems, but relying on the module caching behaviour has bitten us in the past.

Reply:
I agree it's a semi-complex situation with several edge cases, but we should be able to rely on each module being a single instance no matter where we're requiring it from, as that's generally the expected behaviour.

If something else happens, that's a bug we should aim to fix. I would guess the separate builds we use for Riot and SDKs today could cause modules to resolve to different absolute paths, and then trigger separate instances.

In my opinion, a well-functioning module system with clear semantics is important for developer efficiency, so I'd be up for improving this behaviour and adding tests to ensure it works as expected. I filed https://github.com/vector-im/riot-web/issues/9093 to investigate / improve this.